lib/python-sdk/common_grants_sdk/schemas/models/opp_transformer.py

```diff
@@ -0,0 +1,98 @@
+from abc import ABC, abstractmethod
+from datetime import datetime, UTC
+from typing import Annotated, Any, final
+from uuid import UUID, uuid4
+
+from pydantic import Field
+
+from common_grants_sdk.schemas.base import CommonGrantsBaseModel
+from common_grants_sdk.schemas.models.opp_base import OpportunityBase
+from common_grants_sdk.schemas.models.opp_funding import OppFunding
+from common_grants_sdk.schemas.models.opp_status import OppStatus
+from common_grants_sdk.schemas.models.opp_timeline import OppTimeline
+
+
+class OpportunityTransformer(CommonGrantsBaseModel, ABC):
+    """
+    Base class for transforming arbitrary data structures
+    into canonical models.
+    """
+
+    source_data: Annotated[
+        dict[str, Any],
+        Field(
+            description="Arbitrary source data",
+        ),
+    ]
+
+    def transform_opportunity(self, id: UUID | None = None) -> OpportunityBase:
```

Comment:
Related to my comments below on the `self._generate_id()` method, I think it would make more sense to have `id` be a required value that is either a uuid or a string that is used to generate a UUID so that successive runs of this function with the same input data will yield identical results.

Reply:
Or if we're sticking with the pattern of having one instance of this transformer per record, expecting an ID value on initialization or generating and saving the UUID when the transformer is instantiated with input data.